### PR TITLE
ddl: background reorganization change for `admin pause/resume ...`

### PR DESCRIPTION
**Pull request by dhysum**
What problem does this PR solve?
Issue Number: close https://github.com/pingcap/tidb/issues/18015 https://github.com/pingcap/tidb/issues/40041

Related issue number: https://github.com/pingcap/tidb/issues/39751 https://github.com/pingcap/tidb/pull/43061

Problem Summary:  support `admin pause ddl jobs 3,5;` and `admin resume ddl jobs 3,5;`

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [x] Affects user behaviors
- [x] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by zimulala** [comment]
This PR seems that it's duplicated with https://github.com/pingcap/tidb/pull/43171 end https://github.com/pingcap/tidb/pull/43081. Could we close this PR?

File: ddl/ddl_worker.go [hunk 1]
```diff
@@ -952,15 +954,21 @@ func (w *worker) runDDLJob(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64,
 	}()
 	if job.IsFinished() {
 		logutil.Logger(w.logCtx).Debug("[ddl] finish DDL job", zap.String("job", job.String()))
-		return
+		return 0, nil
```
Comment (Benjamin2037) [inline ddl/ddl_worker.go:957]: This changed semantic, before return ver, err,  Now is 0, nil.

File: ddl/ddl_worker.go [hunk 2]
```diff
@@ -952,15 +954,21 @@ func (w *worker) runDDLJob(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64,
...
 	}
+
+	if job.IsPaused() || job.IsPausing() {
+		logutil.Logger(w.logCtx).Info("[ddl] DDL job paused", zap.String("job", job.String()))
+		return ver, pauseReorgWorkers(w, d, job)
```
Comment (Benjamin2037) [inline ddl/ddl_worker.go:962]: Here if job paused or during pausing period, here will return a pause error, but     in caller of runDDLJob func HandleDDLJobTable did not handle the error.

File: ddl/ddl_worker.go [hunk 3]
```diff
@@ -1361,7 +1369,7 @@ func updateSchemaVersion(d *ddlCtx, t *meta.Meta, job *model.Job, multiInfos ...
 			if droppedIDs, ok := job.CtxVars[0].([]int64); ok {
 				if addedIDs, ok := job.CtxVars[1].([]int64); ok {
 					// to use AffectedOpts we need both new and old to have the same length
-					maxParts := mathutil.Max[int](len(droppedIDs), len(addedIDs))
+					maxParts := mathutil.Max(len(droppedIDs), len(addedIDs))
```
Comment (Benjamin2037) [inline ddl/ddl_worker.go:1372]: Seem this line code is no related with pause|resume problem? please do not introduce non-related code change.

File: ddl/reorg.go [hunk 4]
```diff
@@ -228,14 +238,16 @@ func (w *worker) runReorgJob(rh *reorgHandler, reorgInfo *reorgInfo, tblInfo *mo
...
+
+		updateBackfillProgress(w, reorgInfo, tblInfo, rowCount)
+
 		// For other errors, even err is not nil here, we still wait the partial counts to be collected.
 		// since in the next round, the startKey is brand new which is stored by last time.
 		if err != nil {
 			return errors.Trace(err)
 		}
-
-		updateBackfillProgress(w, reorgInfo, tblInfo, 0)
```
Comment (tangenta) [inline ddl/reorg.go:238]: Why is it changed?

Comment (dhysum) [reply ddl/reorg.go:238]: To update the progress if 'dbterror.ErrPausedDDLJob' received.

Even other error may happen, I think it is better to update the progress as well.

File: ddl/ddl_worker.go (hunk 1 above)
Comment (dhysum) [reply ddl/ddl_worker.go:957]: ok

File: ddl/ddl_worker.go (hunk 2 above)
Comment (dhysum) [reply ddl/ddl_worker.go:962]: If the job has been paused, just wait for GetWaitTimeWhenErrorOccurred() and then retry it.

**Review by tangenta — APPROVED** [review]

**Review by Benjamin2037 — APPROVED** [review]
LGTM

**Review by zimulala — COMMENTED** [review]
Please add this test(https://github.com/pingcap/tidb/pull/43171#discussion_r1188067017) here ？

**Comment by dhysum** [comment]
> Please add this test([#43171 (comment)](https://github.com/pingcap/tidb/pull/43171#discussion_r1188067017)) here ？

OK. There should be such cases on the resume action after paused in the cases. Such cases has been done manually, and will add it in another PR later on.

**Pull request merged**